pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/tenant/DefaultTenantRebalancer.java

```diff
@@ -0,0 +1,193 @@
+        config.setProperty(RebalanceConfigConstants.DRY_RUN, true);
+        rebalanceResult.put(table, _pinotHelixResourceManager.rebalanceTable(table, config, false));
+      } catch (TableNotFoundException exception) {
+        rebalanceResult.put(table, new RebalanceResult(null, RebalanceResult.Status.FAILED, exception.getMessage(),
+            null, null, null));
+      }
+    });
+    if (context.isDryRun() || context.isDowntime()) {
+      return new TenantRebalanceResult(null, rebalanceResult, context.isVerboseResult());
+    } else {
+      for (String table : rebalanceResult.keySet()) {
+        RebalanceResult result = rebalanceResult.get(table);
+        if (result.getStatus() == RebalanceResult.Status.DONE) {
+          rebalanceResult.put(table, new RebalanceResult(result.getJobId(), RebalanceResult.Status.IN_PROGRESS,
+              "In progress, check controller task status for the", result.getInstanceAssignment(),
+              result.getTierInstanceAssignment(), result.getSegmentAssignment()));
+        }
+      }
+    }
+
+    String tenantRebalanceJobId = createUniqueRebalanceJobIdentifier();
+    TenantRebalanceObserver observer = new ZkBasedTenantRebalanceObserver(tenantRebalanceJobId, context.getTenantName(),
+        tables, _pinotHelixResourceManager);
+    observer.onTrigger(TenantRebalanceObserver.Trigger.START_TRIGGER, null, null);
+    Set<String> parallelTables;
+    Set<String> sequentialTables;
+    final ConcurrentLinkedQueue<String> parallelQueue;
+    try {
+      if (context.getDegreeOfParallelism() > 1) {
+        if (!context.getParallelWhitelist().isEmpty()) {
+          parallelTables = new HashSet<>(context.getParallelWhitelist());
+        } else {
+          parallelTables = new HashSet<>(tables);
+        }
+        if (!context.getParallelBlacklist().isEmpty()) {
+          parallelTables = Sets.difference(parallelTables, context.getParallelBlacklist());
+        }
+        parallelQueue = new ConcurrentLinkedQueue<>(parallelTables);
+        for (int i = 0; i < context.getDegreeOfParallelism(); i++) {
+          _executorService.submit(() -> {
+            try {
+              while (true) {
+                String table = parallelQueue.remove();
+                Configuration config = extractRebalanceConfig(context);
+                config.setProperty(RebalanceConfigConstants.DRY_RUN, false);
+                config.setProperty(RebalanceConfigConstants.JOB_ID, rebalanceResult.get(table).getJobId());
+                rebalanceTable(table, config, observer);
+              }
+            } catch (NoSuchElementException ignore) {
+            }
+          });
+        }
+        sequentialTables = Sets.difference(tables, parallelTables);
+      } else {
+        sequentialTables = new HashSet<>(tables);
+        parallelQueue = new ConcurrentLinkedQueue<>();
+      }
+      _executorService.submit(() -> {
+        while (!parallelQueue.isEmpty()) {
+          try {
+            Thread.sleep(5000);
```

Comment:
This will hog a thread from the shared executorService. We should yield thread for other tasks while waiting for some condition

Reply:
Hey @saurabhd336 thanks for pointing this out, I have made changes to avoid the thread hogging. Now the last active thread from the parallel threads will pick up the job to rebalance the sequential tables.